File: 0078-subsets/0078-subsets.py
class Solution(object):
    def sol(self, idx, nums, res, ans):
        if (idx == len(nums)):
            ans.append(res[:])
            return

        res.append(nums[idx])
        self.sol(idx + 1, nums, res, ans)

        res.pop()
        self.sol(idx + 1, nums, res, ans)

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        res = []
        ans = []

        self.sol(0, nums, res, ans)

        return ans

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void subseq(int idx, vector<int> &nums, vector<int> &res,vector<vector<int>> &ans){
        
        if(idx == nums.size()){
            ans.push_back(res);
            return;
        }

        res.push_back(nums[idx]);
        subseq(idx+1 , nums, res, ans);

        res.pop_back();
        subseq(idx+1 , nums, res,ans);

    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> res;

        vector<vector<int>> ans;
        subseq(0, nums, res, ans);

        return ans;        

    }
};